qiskit/transpiler/passes/optimization/optimize_annotated.py

```diff
@@ -148,17 +167,220 @@ def _canonicalize(self, dag) -> Tuple[DAGCircuit, bool]:
-        Recursively applies optimizations to op's definition (or to op.base_op's
-        definition if op is an annotated operation).
-        Returns True if did something.
+        Decomposes a circuit ``A`` into 3 sub-circuits ``P``, ``Q``, ``R`` such that
+        ``A = P -- Q -- R`` and ``R = P^{-1}``.
+
+        This is accomplished by iteratively finding inverse nodes at the front and at the back of the
+        circuit.
         """
 
-        # If op is an annotated operation, we descend into its base_op
-        if isinstance(op, AnnotatedOperation):
-            return self._recursively_process_definitions(op.base_op)
+        front_block = []  # nodes collected from the front of the circuit (aka P)
+        back_block = []  # nodes collected from the back of the circuit (aka R)
+
+        # Stores in- and out- degree for each node. These degrees are computed at the start of this
+        # function and are updated when nodes are collected into front_block or into back_block.
+        in_degree = {}
+        out_degree = {}
+
+        # For each qubit, keep the nodes at the start or at the back of the circuit involving this qubit
+        # (also allowing None and empty values).
+        q_to_f = {}
+        q_to_b = {}
+
+        # keep the set of nodes that have been moved either to front_block or to back_block
+        processed_nodes = set()
+
+        # optimization to only consider qubits involved in nodes at the front or at the back of the
+        # circuit
+        unchecked_qubits = set()
+
+        # optimization to keep pairs of nodes for which the inverse check was performed and the nodes
+        # were found to be not inverse to each other
+        checked_node_pairs = set()
+
+        # compute in- and out- degree for every node
+        # also update information for nodes at the start and at the end of the circuit
+        for node in dag.op_nodes():
+            preds = dag.op_predecessors(node)
+            in_degree[node] = len(preds)
+            if len(preds) == 0:
+                for q in node.qargs:
+                    q_to_f[q] = node
+                    unchecked_qubits.add(q)
+            succs = dag.op_successors(node)
+            out_degree[node] = len(succs)
+            if len(succs) == 0:
+                for q in node.qargs:
+                    q_to_b[q] = node
+                    unchecked_qubits.add(q)
+
+        # iterate while there is a possibility to find more inverse pairs
+        while len(unchecked_qubits) > 0:
+            to_check = unchecked_qubits.copy()
+            unchecked_qubits.clear()
+
+            # For each qubit q, check whether the gate at the front of the circuit that involves q
+            # and the gate at the end of the circuit that involves q are inverse
+            for q in to_check:
+
+                if (fn := q_to_f.get(q, None)) is None:
+                    continue
+                if (bn := q_to_b.get(q, None)) is None:
+                    continue
+
```

Comment:
It's a bit weird to have `fn` here but `front_circuit` later.  Might be good to call it the full `front_node` (if not only because then I don't have to double-take about `fn` being the keyword for functions in Rust lol).

Reply:
renaming (and yes I am now also getting confusing thinking this is a function)